src/extractor/tokenizer.ts

```diff
@@ -0,0 +1,124 @@
+import type { IOnigLib, IGrammar } from 'vscode-textmate';
+
+import { join, extname } from 'path';
+import { readFile } from 'fs/promises';
+
+import { Registry, parseRawGrammar, INITIAL } from 'vscode-textmate';
+import { loadWASM, OnigScanner, OnigString } from 'vscode-oniguruma';
+
+export type Token = {
+  type: string;
+  token: string;
+  startIndex: number;
+  endIndex: number;
+  scopes: string[];
+};
+
+const enum Grammar {
+  TYPESCRIPT = 'source.ts',
+  TYPESCRIPT_TSX = 'source.tsx',
+}
+
+const GrammarFiles: Record<Grammar, string> = {
+  [Grammar.TYPESCRIPT]: join(
+    __dirname,
+    'textmate',
+    'TypeScript.tmLanguage'
+  ),
+  [Grammar.TYPESCRIPT_TSX]: join(
+    __dirname,
+    'textmate',
+    'TypeScriptReact.tmLanguage'
+  ),
+};
+
+let oniguruma: Promise<IOnigLib>;
+let registry: Registry;
+
+async function initializeOniguruma() {
+  const wasmBlobPath = require
+    .resolve('vscode-oniguruma')
+    .replace('main.js', 'onig.wasm');
+  const wasmBlob = await readFile(wasmBlobPath);
+  await loadWASM(wasmBlob);
+
+  return <IOnigLib>{
+    createOnigScanner: (patterns) => new OnigScanner(patterns),
+    createOnigString: (s) => new OnigString(s),
+  };
+}
+
+async function loadGrammar(scope: Grammar) {
+  const file = GrammarFiles[scope];
+  if (!file) return null;
+
+  const grammar = await readFile(file, 'utf8');
+  return parseRawGrammar(grammar);
+}
+
+function extnameToGrammar(extname: string) {
+  switch (extname) {
+    case '.js':
+    case '.ts':
+      return Grammar.TYPESCRIPT;
+    case '.jsx':
+    case '.tsx':
+      return Grammar.TYPESCRIPT_TSX;
+  }
+}
+
+function* tokenize(code: string, grammar: IGrammar) {
+  let stack = INITIAL;
+  let linePtr = 0;
+  for (const line of code.split('\n')) {
+    const res = grammar.tokenizeLine(line, stack);
+
+    for (let i = 0; i < res.tokens.length; i++) {
+      const token = res.tokens[i];
+
+      // Opinionated take: if a token is scope-less, chances are we don't care about it.
+      // Ditching it allows us to reduce complexity from the state machine's POV.
+      if (token.scopes.length === 1) continue
```

Comment:
I would rather use brackets with if statements everywhere. It improves readablity imho.